frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/BaseTraverseAction.java

```diff
@@ -0,0 +1,129 @@
+  protected void onUpdate(AnActionEvent anActionEvent) {
+    super.onUpdate(anActionEvent);
+    val presentation = anActionEvent.getPresentation();
+    if (!presentation.isEnabledAndVisible()) {
+      return;
+    }
+
+    presentation.setDescription(getNonHtmlString("action.GitMachete.BaseTraverseAction.description"));
+
+    val graphTable = getGraphTable(anActionEvent);
+    val repositorySnapshot = graphTable.getGitMacheteRepositorySnapshot();
+    val branchLayout = repositorySnapshot != null ? repositorySnapshot.getBranchLayout() : null;
+
+    if (branchLayout == null || branchLayout.getRootEntries().isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription(getNonHtmlString("action.GitMachete.BaseTraverseAction.description.empty-layout"));
+      return;
+    }
+  }
+
+  @Override
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    val gitRepository = getSelectedGitRepository(anActionEvent);
+    val graphTable = getGraphTable(anActionEvent);
+    val repositorySnapshot = graphTable.getGitMacheteRepositorySnapshot();
+    val branchLayout = repositorySnapshot != null ? repositorySnapshot.getBranchLayout() : null;
+    val project = getProject(anActionEvent);
+
+    boolean yesNoResult = true;
+    if (branchLayout != null && branchLayout.getRootEntries().nonEmpty() && gitRepository != null) {
+      if (PropertiesComponent.getInstance(project).getBoolean(SHOW_TRAVERSE_INFO, /* defaultValue */ true)) {
+
+        val traverseInfoDialog = new InfoDialog(
+            project,
+            getString("action.GitMachete.BaseTraverseAction.dialog.traverse-approval.title"),
+            getString("action.GitMachete.BaseTraverseAction.dialog.traverse-approval.text.HTML"),
+            SHOW_TRAVERSE_INFO,
+            /* myHeight */ 100);
+
+        yesNoResult = traverseInfoDialog.showAndGet();
+      }
+
+      if (yesNoResult) {
+        val entryName = getNameOfBranchUnderAction(anActionEvent);
+        if (entryName != null) {
+          checkoutAndTraverse(gitRepository, graphTable, entryName);
+        } else {
+          LOG.warn("Skipping traverse action because entryName is undefined");
+        }
+      }
+    }
+  }
+
+  private void checkoutAndTraverse(GitRepository gitRepository, BaseEnhancedGraphTable graphTable, String branchName) {
+    log().debug(() -> "Queuing '${branchName}' branch checkout background task");
+
+    @UI Runnable traverseRunnable = () -> traverse(gitRepository, graphTable, branchName);
+    Runnable repositoryRefreshRunnable = () -> graphTable.queueRepositoryUpdateAndModelRefresh(traverseRunnable);
+    Runnable callInAwtLater = () -> ModalityUiUtil.invokeLaterIfNeeded(NON_MODAL, repositoryRefreshRunnable);
+    val gitBrancher = GitBrancher.getInstance(gitRepository.getProject());
+    val repositories = Collections.singletonList(gitRepository);
+    gitBrancher.checkout(/* reference */ branchName, /* detach */ false, repositories, callInAwtLater);
```

Comment:
Hmmm since this `Runnable` is supposed to be executed in EDT, I think that `ModalityUiUtil.invokeLaterIfNeeded` shouldn't be necessary. Also, `queueRepositoryUpdateAndModelRefresh` and hence `repositoryRefreshRunnable` isn't even UI-effecting.

Reply:
![image](https://user-images.githubusercontent.com/19799111/205678385-dc89c80a-8486-4934-9531-589603ee51f4.png)